Examples/Io/EDM4hep/src/EDM4hepUtil.cpp

```diff
@@ -0,0 +1,305 @@
+
+        auto parameters = (m.expander() * m.parameters()).eval();
+
+        to.setTime(parameters[Acts::eBoundTime] / Acts::UnitConstants::ns);
+
+        to.setType(EDM4hepUtil::EDM4HEP_ACTS_POSITION_TYPE);
+        // TODO set uv (which are in global spherical coordinates with r=1)
+        to.setPosition({parameters[Acts::eBoundLoc0],
+                        parameters[Acts::eBoundLoc1],
+                        parameters[Acts::eBoundTime]});
+
+        auto covariance =
+            (m.expander() * m.covariance() * m.expander().transpose()).eval();
+        to.setCovMatrix({
+            (float)covariance(Acts::eBoundLoc0, Acts::eBoundLoc0),
+            (float)covariance(Acts::eBoundLoc1, Acts::eBoundLoc0),
+            (float)covariance(Acts::eBoundLoc1, Acts::eBoundLoc1),
+            0,
+            0,
+            0,
+        });
+
+        if (fromCluster != nullptr) {
+          for (auto& c : fromCluster->channels) {
+            auto toChannel = toClusters.create();
+            to.addToRawHits(toChannel.getObjectID());
+
+            // TODO digitization channel
+
+            // TODO get EDM4hep fixed
+            // misusing some fields to store ACTS specific information
+            // don't ask ...
+            toChannel.setType(c.bin[0]);
+            toChannel.setQuality(c.bin[1]);
+            toChannel.setTime(c.activation);
+          }
+        }
+      },
+      from);
+}
+
+void EDM4hepUtil::writeTrajectory(
+    const Trajectories& from, edm4hep::MutableTrack to, std::size_t fromIndex,
+    const IndexMultimap<ActsFatras::Barcode>& hitParticlesMap) {
+  const auto& multiTrajectory = from.multiTrajectory();
+  auto trajectoryState =
+      Acts::MultiTrajectoryHelpers::trajectoryState(multiTrajectory, fromIndex);
+
+  std::vector<ParticleHitCount> particleHitCount;
+  identifyContributingParticles(hitParticlesMap, from, fromIndex,
+                                particleHitCount);
+  // TODO use particles
+
+  // TODO write track params
+  // auto trackParameters = from.trackParameters(fromIndex);
+
+  to.setChi2(trajectoryState.chi2Sum / trajectoryState.NDF);
+  to.setNdf(trajectoryState.NDF);
+
+  multiTrajectory.visitBackwards(fromIndex, [&](const auto& state) {
+    // we only fill the track states with non-outlier measurement
+    auto typeFlags = state.typeFlags();
+    if (!typeFlags.test(Acts::TrackStateFlag::MeasurementFlag)) {
+      return true;
+    }
+
+    Acts::BoundVector meas = state.projector().transpose() * state.parameters();
+    Acts::BoundMatrix cov =
+        state.projector() * state.covariance() * state.projector().transpose();
+
+    edm4hep::TrackState trackState;
+
+    trackState.D0 = meas[Acts::eBoundLoc0];
+    trackState.Z0 = meas[Acts::eBoundLoc1];
+    trackState.phi = meas[Acts::eBoundPhi];
+    trackState.tanLambda = std::tan(M_PI_2 - meas[Acts::eBoundTheta]);
+    trackState.omega = meas[Acts::eBoundQOverP];  // TODO convert
+    trackState.referencePoint = {0, 0, 0};
```

Comment:
Oh I see this is probably what d0/z0 is relative too. Maybe we can convert to perigee at 0, 0, 0?